src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.Engine
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Composition.Hosting;
+
+    using Microsoft.DocAsCode.Build.Common;
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.Exceptions;
+    using Microsoft.DocAsCode.Plugins;
+
+    public class HandlePostProcessors : IDisposable
+    {
+        private readonly List<PostProcessor> _postProcessors;
+
+        public HandlePostProcessors(CompositionHost container, ImmutableArray<string> postProcessorNames)
+        {
+            _postProcessors = GetPostProcessor(container, postProcessorNames);
+        }
+
+        public void PrepareMetadata(ImmutableDictionary<string, object> metadata)
+        {
+            foreach (var postProcessor in _postProcessors)
+            {
+                using (new LoggerPhaseScope($"Prepare metadata in post processor {postProcessor.ContractName}", false))
+                using (new PerformanceScope($"Prepare metadata in post processor {postProcessor.ContractName}", LogLevel.Verbose))
+                {
+                    metadata = postProcessor.Processor.PrepareMetadata(metadata);
+                    if (metadata == null)
+                    {
+                        throw new DocfxException($"Plugin {postProcessor.ContractName} should not return null metadata");
+                    }
+                }
+            }
+        }
+
+        public void PostProcess(Manifest manifest, string outputFolder)
+        {
+            // post process
+            foreach (var postProcessor in _postProcessors)
+            {
+                using (new LoggerPhaseScope($"Process in post processor {postProcessor.ContractName}", false))
```

Comment:
I think it's better to add a overload method to pass in loglevel

---
In reply to: [93849871](https://github.com/dotnet/docfx/pull/1055#discussion_r93849871) [](ancestors = 93849871,93849816,93849068)